### PR TITLE
Fix and tests for issue #10154 inconsistent behavior with invalid dates

### LINKED ISSUE #10154: to_datetime, inconsistent behavior with invalid dates.
Consider Feb 29 1991 (2291991) and March 32 1991 (3321991), Both are invalid dates.
```
pd.to_datetime(2291991, format="%m%d%Y", coerce=True, exact=True)
```
Returns a TypeError: ValueError: day is out of range for month

While
```
pd.to_datetime(3321991, format="%m%d%Y", coerce=True, exact=True)
```
Returns NaT. Which is what I would expect.

In any case they should return the same error or value

Joris Van den Bossche pointed out on the mailing list:
"It has something to do with the number of the day, only values above 31 convert to NaT, 31 or lower raises the error (eg also 31 April raise error instead of giving NaT)"

### PR DESCRIPTION
**Pull request by vincentdavis**
closes #10154 

This addresses the original issue and another discovered with the test cases. 
Please review the tests to be sure you agree with the assertions.

### CODE REVIEW

File: pandas/tseries/tests/test_timeseries.py [hunk 1]
```diff
@@ -4461,6 +4461,23 @@ def test_second(self):
         self.assertIsInstance(r2, Float64Index)
         tm.assert_index_equal(r1, r2)
 
+class TestDaysInMonth(tm.TestCase):
+    def test_day_not_in_month_coerce_true_NaT(self):
+        self.assertTrue(isnull(to_datetime('2015-02-29', coerce=True)))
```
Comment (jreback) [inline pandas/tseries/tests/test_timeseries.py]: add the issue number as a comment

Comment (jreback) [inline pandas/tseries/tests/test_timeseries.py]: pls add a blank line in between functions

File: pandas/tseries/tools.py [hunk 2]
```diff
@@ -349,7 +350,11 @@ def _convert_listlike(arg, box, format):
     elif com.is_list_like(arg):
         return _convert_listlike(arg, box, format)
 
-    return _convert_listlike(np.array([ arg ]), box, format)[0]
+    try:
+        return _convert_listlike(np.array([ arg ]), box, format)[0]
+    except ValueError as e:
+        if not coerce:
```
Comment (jreback) [inline pandas/tseries/tools.py]: what case does this catch here?

Comment (vincentdavis) [inline pandas/tseries/tools.py]: This test would not pass
self.assertTrue(isnull(to_datetime('2015-02-29', format="%Y-%m-%d", coerce=True)))

Comment (jreback) [inline pandas/tseries/tools.py]: this should be caught in the cython code. .

Comment (vincentdavis) [inline pandas/tseries/tools.py]: Without the try/except this is the error. 
It seems like _convert_listlike() would have to accept coerce as an argument to fix it?
```
Traceback (most recent call last):
  File "/Users/vmd/Github/pandas_to_datetime_fix/pandas/tseries/tests/test_timeseries.py", line 4467, in test_day_not_in_month_coerce_true_NaT
    self.assertTrue(isnull(to_datetime('2015-02-29', format="%Y-%m-%d", coerce=True)))
  File "/Users/vmd/Github/pandas_to_datetime_fix/pandas/tseries/tools.py", line 354, in to_datetime
    return _convert_listlike(np.array([ arg ]), box, format)[0]
  File "/Users/vmd/Github/pandas_to_datetime_fix/pandas/tseries/tools.py", line 341, in _convert_listlike
    raise e
  File "/Users/vmd/Github/pandas_to_datetime_fix/pandas/tseries/tools.py", line 332, in _convert_listlike
    if com.is_datetime64_dtype(result) and box:
  File "/Users/vmd/Github/pandas_to_datetime_fix/pandas/core/common.py", line 2516, in is_datetime64_dtype
    tipo = _get_dtype_type(arr_or_dtype)
  File "/Users/vmd/Github/pandas_to_datetime_fix/pandas/core/common.py", line 2479, in _get_dtype_type
    raise ValueError('%r is not a dtype' % arr_or_dtype)
ValueError: None is not a dtype

----------------------------------------------------------------------
Ran 3 tests in 0.019s
```

Comment (jreback) [inline pandas/tseries/tools.py]: coerce is already passed to the cython functions 

my point is that this should be cause and coerced at a lower level

Comment (jreback) [inline pandas/tseries/tools.py]: you are catching an incorrect exception (something else is wrong) - it's not an expected exception

Comment (vincentdavis) [inline pandas/tseries/tools.py]: I reversed this, the tests will not pass now. I don't know how to fix it.

File: pandas/tseries/tests/test_timeseries.py (hunk 1 above)
Comment (vincentdavis) [inline pandas/tseries/tests/test_timeseries.py]: Added comments and blank lines. Should I open new issues for these test cases that do not pass?
```
self.assertTrue(isnull(to_datetime('2015-02-29', format="%Y-%m-%d", coerce=True)))
self.assertTrue(isnull(to_datetime('2015-04-31', format="%Y-%m-%d", coerce=True)))
```

File: pandas/tseries/tools.py (hunk 2 above)
Comment (jreback) [inline pandas/tseries/tools.py]: well I'll have a look tomorrow

**Comment by jreback** [comment]
merged via https://github.com/pydata/pandas/commit/5a76b44f89922a3c3c8656fa039c1434e633da57

thanks!

**Pull request closed without merging**